File: src/test/java/nz/co/searchwellington/tagging/PlaceAutoTaggerTest.java
package nz.co.searchwellington.tagging;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.util.HashSet;
import java.util.Set;

import junit.framework.TestCase;
import nz.co.searchwellington.model.Newsitem;
import nz.co.searchwellington.model.NewsitemImpl;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.repositories.TagDAO;

public class PlaceAutoTaggerTest extends TestCase {
    
    private Tag placesTag;
    private Tag aroValleyTag;
    
    private PlaceAutoTagger placeAutoTagger;
    private Newsitem aroValleyNewsitem;    
    private TagDAO tagDAO = mock(TagDAO.class);
        
    @Override
    protected void setUp() throws Exception {
        placesTag =  new Tag(1, "places","Places", null, new HashSet<Tag>(), 0, false, false);       
        aroValleyTag = new Tag(2, "arovalley","Aro Valley", placesTag, new HashSet<Tag>(), 0, false, false);
        placesTag.addChild(aroValleyTag);
        placeAutoTagger = new PlaceAutoTagger(tagDAO);
        when(tagDAO.loadTagByName("places")).thenReturn(placesTag);        
    }
    
    public void testShouldTagNewsitemsWithPlaceTags() throws Exception {
        aroValleyNewsitem = new NewsitemImpl(1, "Test newsitem", null, ".. Student flats in the Aro Valley... Test", null, null, null, null, null);           
        Set<Tag> suggestedTags = placeAutoTagger.suggestTags(aroValleyNewsitem);
		assertTrue(suggestedTags.contains(aroValleyTag));
    }
    
    public void testPlaceAutoTaggingShouldBeCaseInsensitive() throws Exception {
        aroValleyNewsitem = new NewsitemImpl(1, "Test newsitem", null, ".. Student flats in the aro valley... Test", null, null, null, null, null);           
        Set<Tag> suggestedTags = placeAutoTagger.suggestTags(aroValleyNewsitem);
		assertTrue(suggestedTags.contains(aroValleyTag));
    }
   
}

File: src/test/scala/nz/co/searchwellington/tagging/PlaceAutoTaggerTest.scala
package nz.co.searchwellington.tagging

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.model.{Newsitem, Tag}
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.junit.Assert._
import org.junit.{Before, Test}
import org.mockito.Mockito.{mock, when}
import reactivemongo.api.bson.BSONObjectID

import java.util.UUID
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Await, Future}

class PlaceAutoTaggerTest extends ReasonableWaits {

  private val placesTag = Tag(_id = BSONObjectID.generate, id = UUID.randomUUID().toString, name = "places", display_name = "Places")
  private val aroValleyTag = Tag(id = UUID.randomUUID().toString, name = "arovalley", display_name = "Aro Valley", parent = Some(placesTag._id))
  private val islandBayTag = Tag(id = UUID.randomUUID().toString, name = "islandbay", display_name = "Island Bay", parent = Some(placesTag._id))

  private val aroValleyNewsitem = Newsitem(title = "Test newsitem", description = Some(".. Student flats in the Aro Valley... Test"))

  private val mongoRepository = mock(classOf[MongoRepository])

  private val placeAutoTagger = new PlaceAutoTagger(mongoRepository)

  @Before
  def setUp(): Unit = {
    when(mongoRepository.getTagByUrlWords("places")).thenReturn(Future.successful(Some(placesTag)))
    when(mongoRepository.getTagsByParent(placesTag._id)).thenReturn(Future.successful(List(aroValleyTag, islandBayTag)))
    when(mongoRepository.getTagsByParent(aroValleyTag._id)).thenReturn(Future.successful(List.empty))
    when(mongoRepository.getTagsByParent(islandBayTag._id)).thenReturn(Future.successful(List.empty))
  }

  @Test
  def testShouldTagNewsitemsWithPlaceTags(): Unit = {
    val suggestedTags = Await.result(placeAutoTagger.suggestTags(aroValleyNewsitem), TenSeconds)

    assertTrue(suggestedTags.contains(aroValleyTag))
  }

  @Test
  def testPlaceAutoTaggingShouldBeCaseInsensitive(): Unit = {
    val suggestedTags =  Await.result(placeAutoTagger.suggestTags(aroValleyNewsitem), TenSeconds)

    assertTrue(suggestedTags.contains(aroValleyTag))
  }

}
